packages/formik/src/Formik.tsx

```diff
@@ -171,9 +173,46 @@ export function useFormik<Values extends FormikValues = FormikValues>({
     };
   }, []);
 
+  /**
+   * This is the true test of spacetime. Every method
+   * Formik uses must carefully consider whether it
+   * needs to use the ref or the render snapshot.
+   *
+   * The general rule is going to be,
+   *       snapshot    ref
+   * const [state, updateState] = useFormikThing();
+   */
+  const stateRef = React.useRef<FormikState<Values>>({
+    values: initialValues.current,
+    errors: initialErrors.current,
+    touched: initialTouched.current,
+    status: initialStatus.current,
+    isSubmitting: false,
+    isValidating: false,
+    submitCount: 0,
+  });
+
+  /**
+   * Breaking all the rules, re: "must be side-effect free"
+   * BUT that's probably OK.
+   *
+   * The only things that should use stateRef are side effects themselves --
+   * those things which need the latest value in order to compute their own latest value.
+   */
+  const refBoundFormikReducer = React.useCallback((
```

Comment:
I wonder if this is safe in concurrent mode? I mean it’s unlikely people are suspending in forms, but it’s not impossible. For example, you might suspend to lazy load a date picker or something. Again, above my understanding of CM, but something we should ask React core team about

Reply:
good question. I have no idea, I'll have to look and see how CM works whenever I get some free time to look at this again. From what I remember, I think it means a ref could be updated but then that render abandoned, causing tearing between the two states?? which would definitely not be good for this technique.

Reply:
I had a look at Concurrent Mode and this technique should work as long as we never call `refBoundFormikReducer` outside of Effects or Setters.

What this means, however, is that users can break this pattern by doing something like:

```
const MyForm = () => {
  const { values, setValues } = useFormik(formikProps);
  const dependentValues = calculateDependentValues(values);

  // uh oh, if a render is abandoned after calculating dependentValues 
  // because someone typed another letter, this will be incorrect
  setValues(dependentValues);
  // instead, this MUST use `useEffect`.
  useEffect(() => {
    const dependentValues = calculateDependentValues(values);

    // never incorrect because this render MUST have occurred
    setValues(dependentValues);
  }, [values, setValues]);
}
```

So while it's technically incorrect, it's a _super breaking change_.

Reply:
I try to think of it this way. User interactions should _never_ be abandoned. If a user types a, then b, both value = a and value = ab should be committed. And in this scenario they will.

However, the validation for "a" _can_ be abandoned, and an effect calculating dependent values for "a" can be abandoned, simply by encapsulating useEffect, and that would still work in this scenario.

What differs in this scenario is that the validation for "a", triggered by onChange when validateOnChange is set to true, could actually end up validating "ab", then be followed by a second validation for "ab". I am having trouble reasoning about whether that is more or less correct than the setup we current have.